tetris_3.py

```diff
@@ -75,6 +54,40 @@ def tetris(d):
     return string
 
 
+def update_board_with_piece(array, entry):
+    piece_array = PIECES[entry['piece']]
+    column = entry['column']
+    al = len(piece_array)  # array length
+    # takes part of a piece individually, not good for Z and S
+    while al >= 1:
+        part = piece_array[al - 1]  # start at the last/bottom part
```

Comment:
by treating each row of the piece independently of the rest you're allowing them to potentially 'split' as in the second example (problem with the 'S' piece). the way to fix it would be to run this exact same code, except that instead of immediately setting the row of the piece to the found location on the board (line 73 and following), you defer that until you're done with calculating the row for all the rows of the piece, and then just use the one that further up from the bottom (first obstruction found)